concrete/src/Page/Sitemap/PageListGenerator.php

```diff
@@ -0,0 +1,427 @@
+use Concrete\Core\Entity\Site\Site;
+use Concrete\Core\Multilingual\Page\Section\Section as MultilingualSection;
+use Concrete\Core\Page\Page;
+use Concrete\Core\Permission\Access\Entity\GroupEntity as GroupPermissionAccessEntity;
+use Concrete\Core\Permission\Key\Key as PermissionKey;
+use Concrete\Core\User\Group\Group;
+use DateTime;
+
+/**
+ * Class to be used to generate the list of the pages that should be included in a sitemap.xml file.
+ */
+class PageListGenerator
+{
+    /**
+     * @var \Concrete\Core\Application\Application
+     */
+    protected $app;
+
+    /**
+     * @var \Concrete\Core\Entity\Site\Site|null|false
+     */
+    private $site = false;
+
+    /**
+     * @var \Concrete\Core\Multilingual\Page\Section\Section[]|null
+     */
+    private $multilingualSections = null;
+
+    /**
+     * @var bool|null
+     */
+    private $isMultilingualEnabled = null;
+
+    /**
+     * @var \Concrete\Core\Database\Connection\Connection|null
+     */
+    private $connection = null;
+
+    /**
+     * @var \Concrete\Core\Application\Service\Dashboard|null
+     */
+    private $dashboardHelper = null;
+
+    /**
+     * @var \DateTime|null
+     */
+    private $now = null;
+
+    /**
+     * @var \Concrete\Core\Entity\Attribute\Key\PageKey|null|false
+     */
+    private $excludeFromSiteMapAttributeKey = false;
+
+    /**
+     * @var \Concrete\Core\Permission\Key\PageKey|null|false
+     */
+    private $viewPagePermissionKey = false;
+
+    /**
+     * @var \Concrete\Core\User\Group\Group|null|false
+     */
+    private $visitorsUserGroup = false;
+
+    /**
+     * @var \Concrete\Core\Permission\Access\Entity\GroupEntity|null|false
+     */
+    private $visitorsUserGroupAccessEntity = false;
+
+    /**
+     * Initialize the instance.
+     *
+     * @param Application $app
+     */
+    public function __construct(Application $app)
+    {
+        $this->app = $app;
+    }
+
+    /**
+     * Get the approximage numnber of pages that will be included in the sitemap (the actual value can be lower than it).
+     *
+     * @return int
+     */
+    public function getApproximatePageCount()
+    {
+        $siteTreeIDList = array_merge([0], $this->getSiteTreesIDList());
+        $connection = $this->getConnection();
+        $result = $connection->fetchColumn('select count(*) from Pages where siteTreeID is null or siteTreeID in (' . implode(', ', $siteTreeIDList) . ')');
+
+        return $result ?: 0;
+    }
+
+    /**
+     * Generate the list of pages that should be included in the sitemap.
+     *
+     * @return \Concrete\Core\Page\Page|\Generator
+     */
+    public function generatePageList()
+    {
+        $this->now = new DateTime();
+        $siteTreeIDList = array_merge([0], $this->getSiteTreesIDList());
+        $connection = $this->getConnection();
+        $rs = $connection->executeQuery('select cID from Pages where siteTreeID is null or siteTreeID in (' . implode(', ', $siteTreeIDList) . ')');
```

Comment:
`$this->getConnection()->executeQuery` ?

Reply:
Let's assume that we have this code:

```php
$rs = $this->getConnection()->executeQuery('gino');
```

if tomorrow we'll need to use the connection in another place, we'll have to change those two lines into:

```php
$connection = $this->getConnection();
$rs = $connection->executeQuery('gino');
$rs = $connection->executeQuery('pino');
```

when someone will review the change, he'll need to carefully read the changes to see the relevant changes.
That's why (when I'm not too lazy), I prefer to have a line for every assignment, in particular when the variable is very likely to be reused.

Reply:
Thanks for your explanation. I see it's debatable ;)